File: clients/java/src/main/java/ru/workinprogress/booblik/java/Producer.java
package ru.workinprogress.booblik.java;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Accumulates records and sends them in batches.
 *
 * <p><b>It owns its Connection</b>: one thread holds the pending records and is the only writer to
 * that socket. Do not use the same Connection directly while a Producer has it — responses are
 * matched in order, and a second writer takes somebody else's answer.
 *
 * <p>Records for different partitions accumulate separately and go out as separate requests, because
 * a request addresses one partition — a partition being what has one writer.
 */
public final class Producer implements AutoCloseable {

    /**
     * What a future carries when the batch went out under {@link AckPolicy#NONE}. The record was
     * sent; no offset exists, because none is assigned until the writer reaches the batch.
     */
    public static final long OFFSET_UNKNOWN = -1;

    private final Connection connection;
    private final ProducerConfig config;
    private final BlockingQueue<Command> mailbox = new LinkedBlockingQueue<>();
    private final AtomicBoolean closed = new AtomicBoolean();
    private final Thread loop;

    public Producer(Connection connection, ProducerConfig config) {
        this.connection = connection;
        this.config = config;
        this.loop = new Thread(this::run, "booblik-producer");
        this.loop.setDaemon(true);
        this.loop.start();
    }

    /**
     * Queues a record and returns where its offset will arrive.
     *
     * <p>The record is not on the wire when this returns — that is the point. Await the future to
     * know it landed, or call {@link #flush()} to push everything queued.
     */
    public CompletableFuture<Long> send(String topic, int partition, byte[] record) {
        if (closed.get()) {
            throw new IllegalStateException("booblik: producer is closed");
        }
        CompletableFuture<Long> future = new CompletableFuture<>();
        mailbox.add(new Command(topic, partition, record, future, null));
        return future;
    }

    /** Sends everything queued and waits for the broker to answer all of it. */
    public void flush() {
        CountDownLatch done = new CountDownLatch(1);
        mailbox.add(new Command(null, 0, null, null, done));
        try {
            done.await();
        } catch (InterruptedException interrupted) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("booblik: interrupted while flushing", interrupted);
        }
    }

    /** Flushes what is queued and stops the accumulator. Does not close the Connection. */
    @Override
    public void close() {
        if (!closed.compareAndSet(false, true)) {
            return;
        }
        mailbox.add(Command.CLOSE);
        try {
            loop.join();
        } catch (InterruptedException interrupted) {
            Thread.currentThread().interrupt();
        }
        failLeftovers();
    }

    /**
     * Fails anything that reached the mailbox after the loop had already returned.
     *
     * <p>A {@code send} that passed the closed check just before {@code close} ran can queue its
     * record after the loop is gone, and it would then wait on a future nothing will ever complete.
     * Failing those is the difference between a caller that learns something and a caller that
     * stops.
     *
     * <p>The Go client had the same race with a worse ending — it closed its channel, so the late
     * send panicked outright — and that is how this was noticed at all.
     */
    // Package-private, not private: `ProducerTest` reaches past the public closed check to exercise
    // this path directly. The race it guards against has a window too narrow for threads to hit
    // reliably — a hammering test of it in the Python client passed just as happily with the
    // handling removed — so the honest test is a direct one.
    void failLeftovers() {
        Command leftover;
        while ((leftover = mailbox.poll()) != null) {
            if (leftover.future != null) {
                leftover.future.completeExceptionally(new IllegalStateException("booblik: producer is closed"));
            }
            if (leftover.flush != null) {
                leftover.flush.countDown();
            }
        }
    }

    /** Test-only: queues a record past the public closed check, as a racing caller would. */
    void enqueuePastTheCheck(String topic, int partition, byte[] record, CompletableFuture<Long> future) {
        mailbox.add(new Command(topic, partition, record, future, null));
    }

    private void run() {
        Map<BatchKey, Batch> pending = new HashMap<>();
        long deadline = 0;

        while (true) {
            Command command;
            try {
                if (pending.isEmpty()) {
                    command = mailbox.take();
                } else {
                    long remaining = deadline - System.nanoTime();
                    // `poll` with a timeout either returns an item or null. It cannot take an item
                    // off the queue and then drop it, which is what the equivalent on the JVM's
                    // coroutine client did twice — a timeout there cancels the pending receive, and
                    // a cancelled receive could swallow a record, after which whoever awaited its
                    // offset waited for ever.
                    command = mailbox.poll(Math.max(remaining, 0), TimeUnit.NANOSECONDS);
                }
            } catch (InterruptedException interrupted) {
                Thread.currentThread().interrupt();
                deliver(pending);
                return;
            }

            if (command == null) {
                deliver(pending);
                continue;
            }
            if (command == Command.CLOSE) {
                // Close is a flush: whatever the loop already accepted still goes out. Dropping it
                // would make every clean shutdown a silent data loss.
                deliver(pending);
                return;
            }
            if (command.flush != null) {
                deliver(pending);
                command.flush.countDown();
                continue;
            }

            BatchKey key = new BatchKey(command.topic, command.partition);
            Batch batch = pending.computeIfAbsent(key, ignored -> new Batch());
            batch.records.add(command.record);
            batch.waiting.add(command.future);

            // The window is measured from the **first** record of the batch and never restarted.
            // Timing from the last would let a steady trickle postpone the send indefinitely,
            // turning a latency bound into a latency hope.
            if (pending.size() == 1 && batch.records.size() == 1) {
                deadline = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(config.lingerMillis());
            }
            if (batch.records.size() >= config.maxBatchSize()) {
                deliver(pending);
            }
        }
    }

    private void deliver(Map<BatchKey, Batch> pending) {
        if (pending.isEmpty()) {
            return;
        }

        List<Map.Entry<BatchKey, Batch>> batches = new ArrayList<>(pending.entrySet());
        pending.clear();

        for (Map.Entry<BatchKey, Batch> entry : batches) {
            BatchKey key = entry.getKey();
            Batch batch = entry.getValue();
            try {
                ProduceResult result =
                        connection.produce(key.topic(), key.partition(), batch.records, config.ack());
                for (int index = 0; index < batch.waiting.size(); index++) {
                    // One request is written by one call, so the records are contiguous.
                    batch.waiting
                            .get(index)
                            .complete(result == null ? OFFSET_UNKNOWN : result.baseOffset() + index);
                }
            } catch (RuntimeException failure) {
                // A batch that fails fails for all of its records. Completing some and abandoning
                // the rest would leave callers awaiting a future nothing will ever finish.
                for (CompletableFuture<Long> waiting : batch.waiting) {
                    waiting.completeExceptionally(failure);
                }
            }
        }
    }

    private record BatchKey(String topic, int partition) {}

    private static final class Batch {
        private final List<byte[]> records = new ArrayList<>();
        private final List<CompletableFuture<Long>> waiting = new ArrayList<>();
    }

    private static final class Command {

        private static final Command CLOSE = new Command(null, 0, null, null, null);

        private final String topic;
        private final int partition;
        private final byte[] record;
        private final CompletableFuture<Long> future;
        private final CountDownLatch flush;

        Command(
                String topic,
                int partition,
                byte[] record,
                CompletableFuture<Long> future,
                CountDownLatch flush) {
            this.topic = topic;
            this.partition = partition;
            this.record = record;
            this.future = future;
            this.flush = flush;
        }
    }
}

File: booblik-native/src/nativeMain/kotlin/ru/workinprogress/booblik/native/Producer.kt
package ru.workinprogress.booblik.native

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.launch
import kotlinx.coroutines.newSingleThreadContext
import kotlinx.coroutines.selects.onTimeout
import kotlinx.coroutines.selects.select
import ru.workinprogress.booblik.Offset
import ru.workinprogress.booblik.PartitionId
import ru.workinprogress.booblik.TopicName
import ru.workinprogress.booblik.log.AckPolicy
import ru.workinprogress.booblik.net.wire.ErrorCode
import kotlin.time.TimeSource

/** The broker refused the batch this record was in. */
class ProduceFailedException(
    val code: ErrorCode,
) : IllegalStateException("booblik: broker refused the record: $code")

data class ProducerConfig(
    /** Records per request. Reached first, the batch goes at once. */
    val maxBatchSize: Int = 100,
    /**
     * How long an incomplete batch waits for company, in milliseconds.
     *
     * Zero is not the fast setting. It sends every record on its own, which the broker's own
     * measurements put at 80 592 records/s against 4 335 482 for batches of a hundred.
     */
    val lingerMillis: Long = 5,
    val ackPolicy: AckPolicy = AckPolicy.WRITTEN,
)

/**
 * Accumulates records and sends them in batches.
 *
 * ## Where it runs, and why that took a decision
 *
 * The loop and the socket both live on a thread this producer owns, from `newSingleThreadContext`.
 * That is not one option among several: on Kotlin/Native **`Dispatchers.IO` is `internal`** — the
 * compiler says so, the documentation says otherwise — so there is no IO pool for a blocking socket,
 * and `Dispatchers.Default` has as many threads as the machine has cores. Blocking one of those on
 * a network read is the thing not to do.
 *
 * **[close] must be called.** `newSingleThreadContext` is delicate API for exactly this reason: the
 * thread it starts is not reclaimed on its own, and a producer that is dropped rather than closed
 * leaks it.
 *
 * ## When to use it, measured
 *
 * **Do not await every record through it.** That is the one shape where it loses, and it loses
 * badly: 140 records/s against 595 for the same records sent directly (measurement 23). The reason
 * is arithmetic rather than tuning — a caller waiting for its own offset before sending the next
 * means a batch can never hold more records than there are callers, so the window is paid for a
 * batch that was already as full as it was ever going to get. That is the group-commit finding of
 * measurement 22 in a second place.
 *
 * **Use it when you do not await each record.** There it buys something nothing else here can:
 * **21 692 records/s against 595**, thirty-six times, because [BooblikConnection] is blocking and
 * cannot pipeline — without an accumulator there is no way to have a second record in flight at
 * all. The advantage narrows as callers multiply (2.2× at eight, 1.1× at sixty-four), and what
 * catches up is connections rather than batching: a direct caller has one each and this has one
 * for everybody.
 *
 * If you await every offset, [BooblikConnection.produce] with a list is the better call.
 *
 * Records for different partitions accumulate separately and go out as separate requests, because
 * a request addresses one partition — a partition being what has one writer.
 */
@OptIn(DelicateCoroutinesApi::class, ExperimentalCoroutinesApi::class)
class Producer(
    private val connection: BooblikConnection,
    scope: CoroutineScope,
    private val config: ProducerConfig = ProducerConfig(),
) : AutoCloseable {
    private val dispatcher = newSingleThreadContext("booblik-producer")
    private val mailbox = Channel<Command>(Channel.BUFFERED)

    @Suppress("unused")
    private val job: Job =
        scope.launch(dispatcher) {
            try {
                runLoop()
            } finally {
                drainPending()
            }
        }

    private val pending = LinkedHashMap<Key, Batch>()

    /**
     * Queues [record] and returns a handle that completes with the offset it got.
     *
     * The record is not on the wire when this returns — that is the point. Await the result to know
     * it landed, or [flush] to push everything queued.
     */
    suspend fun send(
        topic: TopicName,
        partition: PartitionId,
        record: ByteArray,
    ): CompletableDeferred<Offset?> {
        val answer = CompletableDeferred<Offset?>()
        mailbox.send(Command.Append(Key(topic, partition), record, answer))
        return answer
    }

    /** Sends everything queued and waits for the broker to answer all of it. */
    suspend fun flush() {
        val done = CompletableDeferred<Unit>()
        mailbox.send(Command.Flush(done))
        done.await()
    }

    /**
     * Stops the accumulator and releases its thread.
     *
     * Not optional. The thread from `newSingleThreadContext` outlives a dropped producer, and a
     * program that creates one per request would run out of threads rather than slow down.
     */
    override fun close() {
        mailbox.close()
        dispatcher.close()
    }

    private suspend fun runLoop() {
        for (first in mailbox) {
            when (first) {
                is Command.Flush -> {
                    sendAll()
                    first.done.complete(Unit)
                    continue
                }

                is Command.Append -> {
                    accumulate(first)
                }
            }

            // The window starts at the first record of the batch. Restarting it on every arrival
            // would let a steady stream defer the send for ever, turning a latency bound into a
            // latency hope.
            val started = TimeSource.Monotonic.markNow()
            while (!isAnyBatchFull()) {
                val remaining = config.lingerMillis - started.elapsedNow().inWholeMilliseconds
                if (remaining <= 0) break

                // `select`, and **not** `withTimeoutOrNull(mailbox.receive())`. A timeout cancels
                // the receive, and a cancelled receive can take an element off the channel and then
                // drop it — after which the caller waiting on that record waits for ever while the
                // accumulator carries on serving everybody else. The broker's own writer had this
                // bug, and so did the JVM client, twice. `select` either takes the element or takes
                // the timeout; never both, never neither.
                val next =
                    select<Command?> {
                        mailbox.onReceiveCatching { it.getOrNull() }
                        onTimeout(remaining) { null }
                    } ?: break

                when (next) {
                    is Command.Append -> {
                        accumulate(next)
                    }

                    is Command.Flush -> {
                        sendAll()
                        next.done.complete(Unit)
                        break
                    }
                }
            }
            sendAll()
        }
    }

    private fun accumulate(command: Command.Append) {
        pending.getOrPut(command.key) { Batch() }.add(command.record, command.answer)
    }

    private fun isAnyBatchFull() = pending.values.any { it.records.size >= config.maxBatchSize }

    private fun sendAll() {
        if (pending.isEmpty()) return
        // `map`, not `entries.toList()`. The latter hands back **views** into the map, and reading
        // `key` off one after `clear()` is undefined — Kotlin/Native's HashMap notices and throws
        // ConcurrentModificationException, while the JVM's LinkedHashMap does not check and returns
        // the stale key, so the same line worked here by accident of one implementation for as long
        // as there was only one. Found by the Kotlin/Native probe in M-134а.
        val batches = pending.map { (key, batch) -> key to batch }
        pending.clear()
        for ((key, batch) in batches) deliver(key, batch)
    }

    private fun deliver(
        key: Key,
        batch: Batch,
    ) {
        val result =
            try {
                connection.produce(key.topic, key.partition, batch.records, config.ackPolicy)
            } catch (failure: Throwable) {
                // A batch that fails fails for all of its records. Completing some and abandoning
                // the rest would leave callers awaiting a result nothing will ever deliver.
                batch.answers.forEach { it.completeExceptionally(failure) }
                return
            }

        if (result == null) {
            // AckPolicy.NONE: sent, and there is no offset to report. Null and not a placeholder —
            // `Offset` refuses negative values, so the other clients' -1 has no equivalent here, and
            // completing with zero would hand back something indistinguishable from the first
            // record of a topic.
            batch.answers.forEach { it.complete(null) }
            return
        }
        if (result.error != ErrorCode.NONE) {
            batch.answers.forEach { it.completeExceptionally(ProduceFailedException(result.error)) }
            return
        }
        // One request is written by one call, so the records are contiguous from the base.
        batch.answers.forEachIndexed { index, answer -> answer.complete(result.baseOffset + index.toLong()) }
    }

    /** Whatever was queued when the mailbox closed still goes out; dropping it would be silent loss. */
    private fun drainPending() {
        sendAll()
        while (true) {
            val leftover = mailbox.tryReceive().getOrNull() ?: break
            when (leftover) {
                is Command.Append -> {
                    leftover.answer.completeExceptionally(
                        IllegalStateException("booblik: producer is closed"),
                    )
                }

                is Command.Flush -> {
                    leftover.done.complete(Unit)
                }
            }
        }
    }

    private data class Key(
        val topic: TopicName,
        val partition: PartitionId,
    )

    private class Batch {
        val records = ArrayList<ByteArray>()
        val answers = ArrayList<CompletableDeferred<Offset?>>()

        fun add(
            record: ByteArray,
            answer: CompletableDeferred<Offset?>,
        ) {
            records += record
            answers += answer
        }
    }

    private sealed interface Command {
        class Append(
            val key: Key,
            val record: ByteArray,
            val answer: CompletableDeferred<Offset?>,
        ) : Command

        class Flush(
            val done: CompletableDeferred<Unit>,
        ) : Command
    }
}
